sympy/simplify/cse_diff.py

```diff
@@ -0,0 +1,171 @@
+            _ = add_to_cache(node)
+    for reduced_expr in reduced_exprs:
+        for node in reduced_expr:
+            _ = add_to_cache(node)
+
+    reduced_matrix = ImmutableDenseMatrix(reduced_exprs).xreplace(expr_to_replacement_cache)
+    replacements = list(replacement_to_reduced_expr_cache.items())
+
+    absolute_derivative_mapping = {}
+    for i, wrt_symbol in enumerate(wrt.args[2]):
+        absolute_derivative = [S.Zero] * len(wrt)
+        absolute_derivative[i] = S.One
+        absolute_derivative_mapping[wrt_symbol] = ImmutableDenseMatrix([absolute_derivative])
+
+    zeros = ImmutableDenseMatrix.zeros(1, len(wrt))
+    for symbol, subexpr in replacements:
+        free_symbols = subexpr.free_symbols
+        absolute_derivative = zeros
+        for free_symbol in free_symbols:
+            replacement_symbol, partial_derivative = add_to_cache(subexpr.diff(free_symbol))
+            absolute_derivative += partial_derivative * absolute_derivative_mapping.get(free_symbol, zeros)
+        absolute_derivative_mapping[symbol] = ImmutableDenseMatrix([[add_to_cache(a)[0] for a in absolute_derivative]])
+
+    replaced_jacobian = ImmutableDenseMatrix.vstack(*[absolute_derivative_mapping.get(e, ImmutableDenseMatrix.zeros(*wrt.shape).T) for e in reduced_matrix])
+
+    required_replacement_symbols = set()
+    stack = [entry for entry in replaced_jacobian if entry.free_symbols]
+    while stack:
+        entry = stack.pop()
+        if entry in required_replacement_symbols or entry in wrt:
+            continue
+        children = list(replacement_to_reduced_expr_cache.get(entry, entry).free_symbols)
+        for child in children:
+            if child not in required_replacement_symbols and child not in wrt:
+                stack.append(child)
+        required_replacement_symbols.add(entry)
+
+    required_replacements_dense = {
+        replacement_symbol: replaced_subexpr
+        for replacement_symbol, replaced_subexpr in replacement_to_reduced_expr_cache.items()
+        if replacement_symbol in required_replacement_symbols
+    }
+
+    counter = Counter(replaced_jacobian.free_symbols)
+    for replaced_subexpr in required_replacements_dense.values():
+        counter.update(replaced_subexpr.free_symbols)
+
+    required_replacements = {}
+    unrequired_replacements = {}
+    for replacement_symbol, replaced_subexpr in required_replacements_dense.items():
+        if isinstance(replaced_subexpr, Symbol) or counter[replacement_symbol] == 1:
+            unrequired_replacements[replacement_symbol] = replaced_subexpr.xreplace(unrequired_replacements)
+        else:
+            required_replacements[replacement_symbol] = replaced_subexpr.xreplace(unrequired_replacements)
+
+    cse_expr = CseExpr((list(required_replacements.items()), replaced_jacobian.xreplace(unrequired_replacements)))
+    if as_cse_expr:
+        return cse_expr
+    return cse_expr.reduced_exprs.subs(reversed(cse_expr.replacements))
```

Comment:
I think CseExpr should have a method that does this like `cse_expr.rebuild()` or something. Also then it isn't really necessary to have the `as_cse_expr` parameter because the caller can just call `.rebuild()` if they want.